apis/pinned-posts/pinned-posts.controller.ts

```diff
@@ -0,0 +1,101 @@
+import * as express from "express";
+import asyncHandler from "express-async-handler";
+import { TypeHelper } from "../../helpers/typeHelper";
+import BitcloutAPI from "../../lib/bitclout/bitclout";
+import JWTAuthMiddleware from "../../middleware/jwt-middleware"
+import { PinnedPostsManager } from "./pinned-posts-manager";
+
+class PinnedPostsController {
+    public router = express.Router();
+
+    private jwtAuthMiddleWare = new JWTAuthMiddleware();
+    private bitclout = new BitcloutAPI();
+    private pinnedPostsManager = new PinnedPostsManager();
+
+    constructor() {
+        this.initializeRoutes();
+    }
+
+    public initializeRoutes() {
+        this.router.post("/pin", this.jwtAuthMiddleWare.protected, asyncHandler(this.pinPost));
+        this.router.post("/unpin", this.jwtAuthMiddleWare.protected, asyncHandler(this.unpinPost));
+        this.router.get("/:publicKey", this.jwtAuthMiddleWare.protected, asyncHandler(this.getPinnedPost));
+    }
+
+    pinPost = async (
+        request: express.Request,
+        response: express.Response
+    ) => {
+        const { postHashHex } = request.body;
+        const publicKey = request.authenticatedPublicKey;
+
+        if (!TypeHelper.isString(publicKey) || !TypeHelper.isString(postHashHex)) {
+            return response.status(400).send({ error: "Requires postHashHex and publicKey" });
+        }
+
+        try {
+            const responses = await Promise.all(
```

Comment:
This ~20 lines is duplicative with the saved posts controller.

Is there a way to make validation of input posts and users more reusable / shared?

e.g. could we make the validators return an error string, or nil and then do something like this? (at least this way we only write error messages in a single place and don't duplicate the logic?)

```
const error = await Promise.all([
                      this.bitclout.errorsForPublicKeyValidation(publicKey),
                      this.bitclout.errorsForPostHashHexValidation(postHashHex)
    ])
    .then((errors) => {  
        return errors.find(err => err != undefined ) 
    });

if (!!error) {
    return response.status(400).send({ success: false, error: error });
}
```

Not in love with it, mostly just thinking out loud

Reply:
I think we broke down the important shared code in the bitcloutAPI. The duplicate code is just the checking for the error. I find it okay if you are okay with it.